AMM/SPECIFICATION.md

```diff
@@ -0,0 +1,180 @@
+
+#### `initialize()`
+
+1. Allows specifying the legitimate exchange contract implementation that the AMM will operate with Requires
+    1. The bytecode root of the desired exchange contract implementation
+
+#### `add_pool()`
+
+1. Allows an asset pair to be registered to the AMM 
+    1. If the AMM is initialized Requires
+    2. Requires
+        1. The identifiers of the two assets
+        2. The exchange contract identifier that is also the identifier of the liquidity pool asset for the given pair if the exchange contract
+            1. Is legitimate, i.e., the bytecode root matches
+            2. Defines the pool for the provided asset pair
+
+### AMM State Checks
+
+#### `pool()`
+
+1. Returns the registered exchange contract identifier for an asset pair Requires
+    1. The identifiers of the two assets
+
+### Exchange Core Functionality
+
+#### `constructor()`
+
+1. Allows specifying the asset pair for which the pool will be managed through the exchange contract 
+    1. If the asset pair for the exchange contract has not already been set 
+    2. Requires two different asset identifiers
+
+#### `deposit()`
+
+1. Allows depositing an asset to the contract (to later be used for adding liquidity or be withdrawn) 
+    1. If the asset pair of the pool is set 
+    2. Requires any amount of either asset in the pair
+
+#### `add_liquidity()`
+
+1. Allows minting liquidity pool assets by using up the deposits when possible 
+    1. If
+        1. The asset pair of the pool is set
+        2. The deadline has not passed
+        3. Deposit amounts of both assets are sufficient
+        4. The resulting liquidity amount is more than the minimum liquidity for a pool to exist. 
+            > **NOTE** This lower limit for liquidity can be modified as desired per asset pair and exists to protect against excessive slippage.
+    2. Requires 
+        1. The desired liquidity amount
+        2. A deadline, i.e., block height limit
+
+#### `remove_liquidity()`
+
+1. Allows burning liquidity pool assets Requires 
+    1. If
+        1. The asset pair of the pool is set
+        2. The deadline has not passed
+        3. The amount of liquidity pool assets are sufficient
+    2. Requires
+        1. The minimum amounts of both assets to receive after burning
+        2. A deadline, i.e., block height limit
+
+#### `withdraw()`
+
+1. Allows withdrawing previously deposited assets without using them to add liquidity 
+    1. If
+        1. The asset pair of the pool is set
+        2. The deposited amount of the asset is sufficient
+    2. Requires an amount of either asset to withdraw
+
+#### `swap_with_exact_input()`
+
+1. Allows selling an exact amount of an asset for the other asset 
+    1. If 
+        1. The asset pair of the pool is set
+        2. The deadline has not passed
+        3. The asset reserve of asset to buy in the contract is sufficient
+        4. The bought amount is not less than the optional minimum desired output amount
```

Comment:
optimal... minimum... desired... output amount. That's a hell of a loaded sentence. What do all of these terms mean? Could you simplify / rephrase so that the intent behind each word is clear?

Reply:
There is no optimal, there is optional. This means that setting this value is not necessary, but will protect against extra slippage if it is set. The minimum is used because it literally means minimum, the swap can result in more. I removed the `desired`.